storage/qingstor.go

```diff
@@ -214,3 +215,85 @@ func (q *QingStorObjectStorage) DeleteObject(objectKey string) (err error) {
 	}
 	return nil
 }
+
+// ListBuckets will list all buckets of the user.
+func (q *QingStorObjectStorage) ListBuckets(zone string) (buckets []string, err error) {
+	res, err := q.service.ListBuckets(&service.ListBucketsInput{Location: convert.String(zone)})
+	if err != nil {
+		log.Errorf("List bucket failed [%v]", err)
+		return nil, err
+	}
+	log.Debugf("<%d> buckets found.\n", *res.Count)
+	for _, b := range res.Buckets {
+		log.Debugf("Bucket <%s>, url <%s>, created <%s>, location <%s>\n",
+			*b.Name, *b.URL, *b.Created, *b.Location)
+		buckets = append(buckets, *b.Name)
+	}
+	return buckets, nil
+}
+
+// ListObjects will list all objects with specific prefix and delimiter from a bucket.
+func (q *QingStorObjectStorage) ListObjects(prefix, delimiter string, marker *string) (oms []*ObjectMeta, err error) {
+	res, err := q.bucket.ListObjects(&service.ListObjectsInput{
+		Delimiter: convert.String(delimiter),
+		Prefix:    convert.String(prefix),
+		Marker:    marker,
+	})
+	if err != nil {
+		log.Errorf("List objects from bucket <%s> at marker <%s> failed [%v]",
+			*q.bucket.Properties.BucketName, convert.StringValue(marker), err)
+		return nil, err
+	}
+	// Add directories into oms (if exists)
+	for _, cpf := range res.CommonPrefixes {
+		oms = append(oms, &ObjectMeta{
+			Key:         convert.StringValue(cpf),
+			ContentType: constants.DirectoryContentType,
+		})
+	}
+	// Add objects into oms
+	for _, obj := range res.Keys {
+		oms = append(oms, &ObjectMeta{
+			convert.StringValue(obj.Key),
+			convert.Int64Value(obj.Size),
+			convert.StringValue(obj.MimeType),
+			convert.StringValue(obj.Etag),
+			time.Unix(int64(convert.IntValue(obj.Modified)), 0),
+			convert.StringValue(obj.StorageClass),
+		})
+	}
+
+	// recursively for next marker request
+	if convert.BoolValue(res.HasMore) {
```

Comment:
Let's use `for` directly, do not call `q.ListObjects`

Reply:
done by next commit